<h1>PROJECT CITYBIKE LUBLIN                               

<span style="color:green">Import all nessesary libraries

In [1]:
import pandas as pd
import numpy as np
import datetime
import matplotlib.pyplot as plt
import plotly.graph_objects as go
%matplotlib inline

<h1>CITYBIKE 2017

<h3> 1.READ DATA

<span style="color:green">Let's read data from 2017.xlsx file.

In [23]:
city_bike2017 = pd.read_excel('2017.xlsx',sheet_name= ['marzec','kwiecień','maj'])

In [3]:
city_bike2017['marzec'].head()

uid  bike_name          start_time            end_time  \
0  26627713      69277 2017-03-21 00:02:18 2017-03-21 00:02:44   
1  26627775      69756 2017-03-21 00:09:40 2017-03-21 00:22:59   
2  26627751      69631 2017-03-21 00:06:02 2017-03-21 00:23:46   
3  26627725      69455 2017-03-21 00:03:41 2017-03-21 00:27:37   
4  26627864      69692 2017-03-21 00:20:58 2017-03-21 00:28:56   

   start_place_id  end_place_id  price  cust_id  
0          315259        315259      0   991786  
1          673845        673845      0   864401  
2          673845        673845      0   864401  
3          315259        692596    100   991786  
4          315248        674263      0   868787

In [4]:
city_bike2017['kwiecień'].head()

uid  bike_name          start_time            end_time  \
0  27188150      69486 2017-03-31 23:56:26 2017-04-01 00:00:12   
1  27187806      69557 2017-03-31 23:45:38 2017-04-01 00:00:39   
2  27187791      69269 2017-03-31 23:45:14 2017-04-01 00:01:13   
3  27187827      69321 2017-03-31 23:46:13 2017-04-01 00:01:24   
4  27187995      69344 2017-03-31 23:51:28 2017-04-01 00:01:55   

   start_place_id  end_place_id  price  cust_id  
0          315273        315273      0  1514820  
1          315272        315269      0  2024593  
2          315272        315269      0  2024593  
3          315272        315269      0  2024593  
4          315248        315233      0   983017

In [5]:
city_bike2017['maj'].head()

uid  bike_name          start_time            end_time  \
0  28587694      69328 2017-04-30 23:57:00 2017-05-01 00:01:00   
1  28587504      69174 2017-04-30 23:43:00 2017-05-01 00:02:00   
2  28587316      69391 2017-04-30 23:31:00 2017-05-01 00:02:00   
3  28587629      69507 2017-04-30 23:51:00 2017-05-01 00:02:00   
4  28587614      69659 2017-04-30 23:50:00 2017-05-01 00:03:00   

   start_place_id  end_place_id  price  cust_id  
0          360123        315269      0  2043725  
1          315235        674353      0   865781  
2          315265        315248    100   975251  
3          315238        315236      0  1300588  
4          315242        315269      0   886601

In [6]:
#You have to merge dataframes by using 'append'

city_bike17 = city_bike2017['marzec'].append(city_bike2017['kwiecień'], ignore_index= True)
city_bike17 = city_bike17.append(city_bike2017['maj'], ignore_index= True)
city_bike17.head()

uid  bike_name          start_time            end_time  \
0  26627713      69277 2017-03-21 00:02:18 2017-03-21 00:02:44   
1  26627775      69756 2017-03-21 00:09:40 2017-03-21 00:22:59   
2  26627751      69631 2017-03-21 00:06:02 2017-03-21 00:23:46   
3  26627725      69455 2017-03-21 00:03:41 2017-03-21 00:27:37   
4  26627864      69692 2017-03-21 00:20:58 2017-03-21 00:28:56   

   start_place_id  end_place_id  price  cust_id  
0          315259        315259      0   991786  
1          673845        673845      0   864401  
2          673845        673845      0   864401  
3          315259        692596    100   991786  
4          315248        674263      0   868787

<span style="color:green">What would you like to  know about data?
    
<span style="color:green">Use DataFrame.info() and get information about:
    
<span style="color:green">-numbers of rows and coulmns 
    
<span style="color:green">-types of data
    
<span style="color:green">-are there missing values? 
    
<span style="color:green">-how many memory we need

In [7]:
city_bike17.info(memory_usage = 'deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 201078 entries, 0 to 201077
Data columns (total 8 columns):
uid               201078 non-null int64
bike_name         201078 non-null int64
start_time        201078 non-null datetime64[ns]
end_time          201078 non-null datetime64[ns]
start_place_id    201078 non-null int64
end_place_id      201078 non-null int64
price             201078 non-null int64
cust_id           201078 non-null int64
dtypes: datetime64[ns](2), int64(6)
memory usage: 12.3 MB


<span style="color:green">There are 201078 rows and 8 columns.

<span style="color:green">Dtypes: datetime64(2), int64(6)

<span style="color:green">There are not missing value - every columns has the same num non-null values

<span style="color:green">We use 13.8MB

In [8]:
#You can also use DataFrame.shape to get num of rows and columns and DataFrame.isnull to get num of non-null values
print(city_bike17.shape )
print(city_bike17.isnull().any().any()) #False means that you don't have any non-missing values'''

(201078, 8)
False


In [9]:
#You can use DataSeries.astype('type') to change memory usage

#TUTAJ COŚ POSZŁO NIE TAK
'''
def memory_usage(df):
    for series in (['start_place_id','end_place_id','price']):
        df[series] = df[series].astype('int16')
    for serie in (['bike_name']):
        df[serie] = df[serie].astype('int32')
    return df
city_bike17=memory_usage(city_bike17)'''
#city_bike17.info(memory_usage='deep')

"\ndef memory_usage(df):\n    for series in (['start_place_id','end_place_id','price']):\n        df[series] = df[series].astype('int16')\n    for serie in (['bike_name']):\n        df[serie] = df[serie].astype('int32')\n    return df\ncity_bike17=memory_usage(city_bike17)"

In [10]:
#create additional TIME columns so it would be easier to manipulate the data

city_bike17['start_day'] = pd.DatetimeIndex(city_bike17['start_time']).dayofyear

city_bike17['start_month'] = pd.DatetimeIndex(city_bike17['start_time']).month

city_bike17['day_of_month'] = pd.DatetimeIndex(city_bike17['start_time']).day

city_bike17['day_of_week'] = pd.DatetimeIndex(city_bike17['start_time']).dayofweek



#create additional duration column so it would be easier to manipulate the data

city_bike17['duration_min']= city_bike17['end_time']-city_bike17['start_time']
#city_bike17['duration_min']=pd.to_datetime(city_bike17['duration_min'])
city_bike17.head()


uid  bike_name          start_time            end_time  \
0  26627713      69277 2017-03-21 00:02:18 2017-03-21 00:02:44   
1  26627775      69756 2017-03-21 00:09:40 2017-03-21 00:22:59   
2  26627751      69631 2017-03-21 00:06:02 2017-03-21 00:23:46   
3  26627725      69455 2017-03-21 00:03:41 2017-03-21 00:27:37   
4  26627864      69692 2017-03-21 00:20:58 2017-03-21 00:28:56   

   start_place_id  end_place_id  price  cust_id  start_day  start_month  \
0          315259        315259      0   991786         80            3   
1          673845        673845      0   864401         80            3   
2          673845        673845      0   864401         80            3   
3          315259        692596    100   991786         80            3   
4          315248        674263      0   868787         80            3   

   day_of_month  day_of_week duration_min  
0            21            1     00:00:26  
1            21            1     00:13:19  
2            21            1     00:17:44  
3            21            1     00:23:56  
4            21            1     00:07:58

In [11]:
#city_bike17.sort_values(by = ['duration_min'],ascending=False).head(1)
#city_bike17.iloc[[60233]]

<h3>2.DATA ANALYSIS

In [12]:
#For how many days were the bikes available for rental?

num_of_days = city_bike17['start_day'].nunique()

print(f'The bikes were available for rental for {num_of_days} dyas')

The bikes were available for rental for 72 dyas


In [13]:
#how many rentals and bikes were there?

num_of_bikes = city_bike17['bike_name'].nunique()

num_of_rent = city_bike17['bike_name'].count()

print(f'Count of rentals:{num_of_rent}, \nCount of bikes: {num_of_bikes}')

Count of rentals:201078, 
Count of bikes: 850


In [14]:
#The average time of a rental:

avg_time = city_bike17['duration_min'].mean()

avg_time.components.minutes

print(f'The average time of rental: {avg_time.components.minutes}')

The average time of rental: 16


In [15]:
#Rentals under 15 minutes


city_bike17['duration_min'] = pd.DatetimeIndex(city_bike17['duration_min']).minute

rentals_under_15 = len(city_bike17[city_bike17['duration_min']<15])

print('Num of rentals under 15 minutes:', rentals_under_15)

Num of rentals under 15 minutes: 137776


C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: FutureWarning:

Passing timedelta64-dtype data is deprecated, will raise a TypeError in a future version



In [16]:
#Number of rentals per month, most popular month


top_month = city_bike17.groupby('start_month').agg({'bike_name':'count'}).sort_values(by='bike_name', ascending=False).iloc[0].name
print(f'The most popular month is:{top_month}')

city_bike17.groupby('start_month').agg({'bike_name':'count'}).sort_values(by=['bike_name'], ascending=False)

The most popular month is:5


bike_name
start_month           
5               105776
4                67152
3                28150

In [17]:
#daily top score

city_bike17['date'] = pd.DatetimeIndex(city_bike17['start_time']).date

print(city_bike17.groupby('date').agg({'bike_name':'count'}).sort_values(by='bike_name', ascending=False).iloc[0].name)

city_bike17.groupby(['start_day']).agg({'bike_name':'count'}).sort_values(by=['bike_name'],ascending=False).head(1)


2017-04-02


bike_name
start_day           
92              6092

In [18]:
#most popular start station

city_bike17['start_place_id'].value_counts().head(1)

315269    9187
Name: start_place_id, dtype: int64

In [19]:
#most popular end station

city_bike17['end_place_id'].value_counts().head(1)

315269    9967
Name: end_place_id, dtype: int64

In [20]:
#most rented bike

city_bike17['bike_name'].value_counts().head(1)

69617    439
Name: bike_name, dtype: int64